File: python/mirascope/llm/providers/openai/completions/_utils/feature_info.py
"""Model feature information for OpenAI completions encoding."""

from __future__ import annotations

from dataclasses import dataclass

from ...model_info import (
    MODELS_WITHOUT_AUDIO_SUPPORT,
    MODELS_WITHOUT_JSON_OBJECT_SUPPORT,
    MODELS_WITHOUT_JSON_SCHEMA_SUPPORT,
    NON_REASONING_MODELS,
)


@dataclass(frozen=True)
class CompletionsModelFeatureInfo:
    """Model feature information for OpenAI completions encoding.

    This dataclass encapsulates feature detection for OpenAI-compatible models,
    allowing providers to pass pre-computed feature information rather than
    relying on model name matching in encode_request.

    None values mean "unknown":
        - audio_support: None → allow audio (permissive)
        - strict_support: None → default to tool mode, but allow explicit strict
        - json_object_support: None → disable (use prompt instructions instead)
        - is_reasoning_model: None → treat as False (allow temperature)
    """

    audio_support: bool | None = None
    """Whether the model supports audio inputs. None means skip check (allow)."""

    strict_support: bool | None = None
    """Whether the model supports strict JSON schema. None allows explicit strict."""

    json_object_support: bool | None = None
    """Whether the model supports JSON object response format. None disables."""

    is_reasoning_model: bool | None = None
    """Whether the model is a reasoning model. None means False (allow temperature)."""


def feature_info_for_openai_model(model_name: str) -> CompletionsModelFeatureInfo:
    """Get feature info for a base OpenAI model name."""
    return CompletionsModelFeatureInfo(
        audio_support=model_name not in MODELS_WITHOUT_AUDIO_SUPPORT,
        strict_support=model_name not in MODELS_WITHOUT_JSON_SCHEMA_SUPPORT,
        json_object_support=model_name not in MODELS_WITHOUT_JSON_OBJECT_SUPPORT,
        is_reasoning_model=model_name not in NON_REASONING_MODELS,
    )

File: typescript/src/llm/providers/openai/completions/_utils/feature-info.ts
/**
 * Model feature information for OpenAI completions encoding.
 */

import {
  MODELS_WITHOUT_AUDIO_SUPPORT,
  MODELS_WITHOUT_JSON_OBJECT_SUPPORT,
  MODELS_WITHOUT_JSON_SCHEMA_SUPPORT,
  NON_REASONING_MODELS,
} from "@/llm/providers/openai/model-info";

/**
 * Model feature information for OpenAI completions encoding.
 *
 * This interface encapsulates feature detection for OpenAI-compatible models,
 * allowing providers to pass pre-computed feature information rather than
 * relying on model name matching in buildRequestParams.
 *
 * Undefined/null values mean "unknown":
 * - audioSupport: undefined → allow audio (permissive)
 * - strictSupport: undefined → default to tool mode, but allow explicit strict
 * - jsonObjectSupport: undefined → disable (use prompt instructions instead)
 * - isReasoningModel: undefined → treat as false (allow temperature)
 */
export interface CompletionsModelFeatureInfo {
  /** Whether the model supports audio inputs. undefined means skip check (allow). */
  readonly audioSupport?: boolean;

  /** Whether the model supports strict JSON schema. undefined allows explicit strict. */
  readonly strictSupport?: boolean;

  /** Whether the model supports JSON object response format. undefined disables. */
  readonly jsonObjectSupport?: boolean;

  /** Whether the model is a reasoning model. undefined means false (allow temperature). */
  readonly isReasoningModel?: boolean;
}

/**
 * Empty feature info with all values undefined.
 * Used as the default for unknown models (most permissive).
 */
export const EMPTY_FEATURE_INFO: CompletionsModelFeatureInfo = Object.freeze(
  {},
);

/**
 * Get feature info for a base OpenAI model name.
 *
 * @param modelName - The base model name (e.g., "gpt-4o", not "openai/gpt-4o")
 * @returns Feature info with all fields populated based on known model capabilities
 */
export function featureInfoForOpenAIModel(
  modelName: string,
): CompletionsModelFeatureInfo {
  return {
    audioSupport: !MODELS_WITHOUT_AUDIO_SUPPORT.has(modelName),
    strictSupport: !MODELS_WITHOUT_JSON_SCHEMA_SUPPORT.has(modelName),
    jsonObjectSupport: !MODELS_WITHOUT_JSON_OBJECT_SUPPORT.has(modelName),
    isReasoningModel: !NON_REASONING_MODELS.has(modelName),
  };
}
